### PR TITLE
py-particle: new versions 0.16.*, 0.20.*, 0.21.*

### PR DESCRIPTION
**Pull request by wdconinc**
https://github.com/scikit-hep/particle/compare/v0.15.1...v0.21.2

Python 2 support was dropped in 0.20.

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-particle/package.py [hunk 1]
```diff
@@ -20,11 +20,22 @@ class PyParticle(PythonPackage):
...
+    version("0.16.2", sha256="53581654ee6b3efbe0993317cff4d83f5529c1da851e12c77a6f6397956100ff")
+    version("0.16.1", sha256="40413c9bae8c42d437cbb280717d9a53b34e496833efea8779eb5f5ef7bfb096")
+    version("0.16.0", sha256="582371c69dd897dea3653ae82b7c2602635e4b797b832e141a856746ba9b0b98")
     version("0.15.1", sha256="6b05cdc4b76c70f785e89258a470504ad87ca119057c65da30a7d4412cca824f")
     version("0.14.1", sha256="05b345f8fbfdb12a0aa744c788b6e1b22326b5a6ad95230596e0fc9ebad56621")
     version("0.11.0", sha256="e90dc36c8b7d7431bd14ee5a28486d28b6c0708555845d1d7bdf59a165405f12")
 
     depends_on("python@2.7:2.8,3.5:", type=("build", "run"))
+    depends_on("python@3.6:", when="@0.20:")
+    depends_on("python@3.7:", when="@0.21:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-particle/package.py]: Spack only supports 3.7+, so you could actually remove all these lines.

If you keep them, you need to fix the dep type

File: var/spack/repos/builtin/packages/py-particle/package.py [hunk 2]
```diff
@@ -20,11 +20,22 @@ class PyParticle(PythonPackage):
...
     depends_on("py-setuptools", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-particle/package.py]: Latest versions use hatchling/hatch-vcs, not setuptools

File: var/spack/repos/builtin/packages/py-particle/package.py [hunk 3]
```diff
@@ -20,11 +20,22 @@ class PyParticle(PythonPackage):
...
 
     depends_on("py-attrs@19.2.0:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-particle/package.py:46]: Missing deps:
```
    "importlib-resources>=2.0;python_version<\"3.9\"",
    "typing-extensions;python_version<\"3.8\"",
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-particle/package.py (hunk 1 above)
Comment (wdconinc) [reply var/spack/repos/builtin/packages/py-particle/package.py]: Kept and dep type fixed.

File: var/spack/repos/builtin/packages/py-particle/package.py (hunk 2 above)
Comment (wdconinc) [reply var/spack/repos/builtin/packages/py-particle/package.py]: Modified to use py-hatchling for newer versions.

File: var/spack/repos/builtin/packages/py-particle/package.py (hunk 3 above)
Comment (wdconinc) [reply var/spack/repos/builtin/packages/py-particle/package.py:46]: Included as dependencies.

File: var/spack/repos/builtin/packages/py-particle/package.py [hunk 4]
```diff
@@ -20,12 +20,27 @@ class PyParticle(PythonPackage):
...
-    depends_on("python@2.7:2.8,3.5:", type=("build", "run"))
-    depends_on("py-setuptools", type="build")
+    depends_on("python@2.7:2.8,3.5:", when="@:0.19", type=("build", "run"))
+    depends_on("python@3.6:", when="@0.20:", type=("build", "run"))
+    depends_on("python@3.7:", when="@0.21:", type=("build", "run"))
+    depends_on("py-setuptools", when="@:0.20", type="build")
+    depends_on("py-hatchling", when="@0.21:", type="build")
+    depends_on("py-hatch-vcs", when="@0.21:", type="build")
+    depends_on("py-importlib-resources@2:", when="^python@:3.8", type=("build", "run"))
+    depends_on("py-typing-extensions", when="^python@:3.7", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-particle/package.py]: This is only in the latest release, older releases required different versions, or didn't require these at all

Comment (wdconinc) [reply var/spack/repos/builtin/packages/py-particle/package.py]:
```suggestion
    depends_on("py-importlib-resources@2:", when="@0.21: ^python@:3.8", type=("build", "run"))
    depends_on("py-typing-extensions", when="@0.21: ^python@:3.7", type=("build", "run"))
```

**Comment by wdconinc** [comment]
After all the suggested and requested changes, the latest version builds fine for all new versions on my system:
```console
17:00:26 wdconinc@menelaos ~ $ spack find -l -v
==> In environment py-particle
==> Root specs
------- py-particle@0.16.0  ------- py-particle@0.16.2  ------- py-particle@0.20.0  ------- py-particle@0.21.0  ------- py-particle@0.21.2
------- py-particle@0.16.1  ------- py-particle@0.16.3  ------- py-particle@0.20.1  ------- py-particle@0.21.1

==> Installed packages
-- linux-ubuntu22.10-skylake / gcc@12.2.0 -----------------------
q5eauf6 berkeley-db@18.1.40+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc
4wb2p57 bzip2@1.0.8~debug~pic+shared build_system=generic
4dxysjc ca-certificates-mozilla@2023-01-10 build_system=generic
zsh6vs3 diffutils@3.8 build_system=autotools
azxc2jf expat@2.5.0+libbsd build_system=autotools
552fqhg gdbm@1.23 build_system=autotools
zntvyct gettext@0.21.1+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools
ikahixk libbsd@0.11.7 build_system=autotools
ao44gct libffi@3.4.4 build_system=autotools
45jm3xu libiconv@1.17 build_system=autotools libs=shared,static
p34w56z libmd@1.0.4 build_system=autotools
wt2rdt5 libxcrypt@4.4.33~obsolete_api build_system=autotools
6axl3as libxml2@2.10.3~python build_system=autotools
dj4mdec ncurses@6.4~symlinks+termlib abi=none build_system=autotools
nlwmceq openssl@1.1.1t~docs~shared build_system=generic certs=mozilla
w7w5ft6 perl@5.36.0+cpanm+open+shared+threads build_system=generic
fezcfme pigz@2.7 build_system=makefile
l7agxyw pkgconf@1.8.0 build_system=autotools
xx5gctu py-attrs@22.1.0 build_system=python_pip
bn6wapv py-editables@0.3 build_system=python_pip
twsdf7t py-flit-core@3.7.1 build_system=python_pip
dlwawdb py-hatch-vcs@0.2.0 build_system=python_pip
753n5lx py-hatchling@1.10.0 build_system=python_pip
hq5kokk py-hepunits@2.3.1 build_system=python_pip
dac2jqx py-packaging@23.0 build_system=python_pip
s6ebvya py-particle@0.16.0 build_system=python_pip
cuj2r37 py-particle@0.16.1 build_system=python_pip
wqs22ht py-particle@0.16.2 build_system=python_pip
auehlfn py-particle@0.16.3 build_system=python_pip
pc7637o py-particle@0.20.0 build_system=python_pip
ik4fazx py-particle@0.20.1 build_system=python_pip
o7grpzk py-particle@0.21.0 build_system=python_pip
4kzjjcz py-particle@0.21.1 build_system=python_pip
s52spro py-particle@0.21.2 build_system=python_pip
x6llv72 py-pathspec@0.10.1 build_system=python_pip
ta5r4ni py-pip@23.0 build_system=generic
augq2mk py-pluggy@1.0.0 build_system=python_pip
rgz7hhb py-setuptools@65.5.0 build_system=generic
ke3eyv3 py-setuptools-scm@7.0.5+toml build_system=python_pip
n5jabin py-toml@0.10.2 build_system=python_pip
e6me3li py-tomli@2.0.1 build_system=python_pip
bvonp6a py-typing-extensions@4.3.0 build_system=python_pip
i76tg6g py-wheel@0.37.1 build_system=generic
ucohqwu python@3.10.8+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,7d40923,f2fd060
u6uj5ti readline@8.2 build_system=autotools patches=bbf97f1
ojb4iwl sqlite@3.40.1+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools
rwvdxr5 tar@1.34 build_system=autotools zip=pigz
veimqlq util-linux-uuid@2.38.1 build_system=autotools
6ocy5dp xz@5.4.1~pic build_system=autotools libs=shared,static
7ppfx4x zlib@1.2.13+optimize+pic+shared build_system=makefile
exhnx5g zstd@1.5.2+programs build_system=makefile compression=none libs=shared,static
==> 51 installed packages
```

File: var/spack/repos/builtin/packages/py-particle/package.py [hunk 5]
```diff
@@ -20,12 +20,27 @@ class PyParticle(PythonPackage):
...
-    depends_on("python@2.7:2.8,3.5:", type=("build", "run"))
-    depends_on("py-setuptools", type="build")
+    depends_on("python@2.7:2.8,3.5:", when="@:0.19", type=("build", "run"))
+    depends_on("python@3.6:", when="@0.20:", type=("build", "run"))
+    depends_on("python@3.7:", when="@0.21:", type=("build", "run"))
+    depends_on("py-setuptools", when="@:0.20", type="build")
+    depends_on("py-hatchling", when="@0.21:", type="build")
+    depends_on("py-hatch-vcs", when="@0.21:", type="build")
+    depends_on("py-importlib-resources@2:", when="@0.21: ^python@:3.8", type=("build", "run"))
+    depends_on("py-typing-extensions", when="@0.21: ^python@:3.7", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-particle/package.py]: I think these are required as far back as 0.16 aren't they?
```suggestion
    depends_on("py-importlib-resources@2:", when="@0.16: ^python@:3.8", type=("build", "run"))
    depends_on("py-typing-extensions", when="@0.16: ^python@:3.7", type=("build", "run"))
```

File: var/spack/repos/builtin/packages/py-particle/package.py [hunk 6]
```diff
@@ -20,12 +20,27 @@ class PyParticle(PythonPackage):
...
     version("0.15.1", sha256="6b05cdc4b76c70f785e89258a470504ad87ca119057c65da30a7d4412cca824f")
     version("0.14.1", sha256="05b345f8fbfdb12a0aa744c788b6e1b22326b5a6ad95230596e0fc9ebad56621")
     version("0.11.0", sha256="e90dc36c8b7d7431bd14ee5a28486d28b6c0708555845d1d7bdf59a165405f12")
 
-    depends_on("python@2.7:2.8,3.5:", type=("build", "run"))
-    depends_on("py-setuptools", type="build")
+    depends_on("python@2.7:2.8,3.5:", when="@:0.19", type=("build", "run"))
+    depends_on("python@3.6:", when="@0.20:", type=("build", "run"))
+    depends_on("python@3.7:", when="@0.21:", type=("build", "run"))
+    depends_on("py-setuptools", when="@:0.20", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-particle/package.py:39]: setuptools-scm is also required in some versions

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-particle/package.py (hunk 5 above)
Comment (wdconinc) [reply var/spack/repos/builtin/packages/py-particle/package.py]: Right, that's in `setup.cfg` for those older versions. Sorry for all the back and forth. Initial forays in python package updates...

File: var/spack/repos/builtin/packages/py-particle/package.py (hunk 6 above)
Comment (wdconinc) [reply var/spack/repos/builtin/packages/py-particle/package.py:39]: So, py-setuptools-scm was added to pyproject.toml in https://github.com/scikit-hep/particle/commit/319865e91f3d858c77822f50c57195d1755629b9, i.e. 0.10.0, so should it have been in here all along?
```suggestion
    depends_on("py-setuptools", when="@:0.20", type="build")
    depends_on("py-setuptools-scm", when="@:0.20", type="build")
```

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-particle/package.py:39]: Yep

File: var/spack/repos/builtin/packages/py-particle/package.py [hunk 7]
```diff
@@ -20,12 +20,28 @@ class PyParticle(PythonPackage):
...
+    depends_on("py-setuptools-scm", when="@:0.20", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-particle/package.py]:
```suggestion
    depends_on("py-setuptools-scm@3.4:+toml", when="@:0.20", type="build")
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
Will give @vvolkl a few days to review this, ping me if it's been a week and you haven't heard back and I'll merge

**Comment by alalazo** [comment]
A week has passed, so merging

**Pull request merged**